orttraining/orttraining/test/gradient/function_ops_test.cc

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <iostream>
+
+#include "gtest/gtest.h"
+#include "core/graph/model.h"
+#include "core/graph/contrib_ops/contrib_defs.h"
+#include "orttraining/core/graph/training_op_defs.h"
+#include "test/test_environment.h"
+
+#include "core/session/inference_session.h"
+#include "core/providers/cpu/cpu_execution_provider.h"
+
+#include "test/framework/test_utils.h"
+
+using namespace ::onnxruntime::common;
+
+namespace onnxruntime {
+namespace test {
+
+typedef std::vector<onnxruntime::NodeArg*> ArgMap;
+
+static ONNX_NAMESPACE::TypeProto TensorType(int32_t elem_type, std::vector<int64_t> dims) {
+  ONNX_NAMESPACE::TypeProto typeProto;
+  typeProto.mutable_tensor_type()->set_elem_type(elem_type);
+  auto* shape = typeProto.mutable_tensor_type()->mutable_shape();
+  for (auto dim : dims)
+    shape->add_dim()->set_dim_value(dim);
+  return typeProto;
+}
```

Comment:
Could we add test case like [batch,seq,512]?

Reply:
There is already a 3D test-case. But it has default axis. Do you want a 3D-case with some other axis-value?

Reply:
I was thinking about symbolic dim vs. static dim. Seems the test case only handles static dims.

---
In reply to: [592862644](https://github.com/microsoft/onnxruntime/pull/6988#discussion_r592862644) [](ancestors = 592862644)